### DISCUSSION TITLE
"sequencer" field of bucket notification event should provide accurate ordering of PUT/DELETE events on the same object

### DISCUSSION THREAD

**Discussion by mctoohey** [minio/minio#18274] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
According to the documentation on [Amazon S3 Event Notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-content-structure.html) the "sequencer" field 
> can be used to determine the order of events for a given object key. 

with the following caveats 
> * You can't use sequencer to determine order for events on different object keys.
> * The sequencers can be of different lengths. So, to compare these values, first right pad the shorter value with zeros, and then do a lexicographical comparison.

Currently MinIO fills this field with a nanosecond timestamp encoded as hex at the time the event is created.

https://github.com/minio/minio/blob/557df666fde93cee12bdf7af4468925d1318e25d/cmd/event-notification.go#L211C2-L212

https://github.com/minio/minio/blob/557df666fde93cee12bdf7af4468925d1318e25d/cmd/event-notification.go#L256C9-L256C9

While this means the "sequencer" field is probably correct most of the time. If two events on the same object happen very close together it is possible that the "sequencer" field could be out of order.

***Why does this matter?***

This is important if you want to be able to use bucket notification events to determine the existence of an object. If a PUT then a DELETE happen on the same object a short time apart. Then the object no longer exists. But if the ordering of events make it look like it was a DELETE then a PUT, observing the events in this order makes it look like the object does exist.

**Describe the solution you'd like**
A nanosecond timestamp may still be able to work for this. As long as the timestamp is taken while the RW mutex on the object is held during the PUT/DELETE operation. And with the assumption that the system clocks on all nodes in the cluster are accurately synced.

**Describe alternatives you've considered**

**Additional context**

**Comment by klauspost** [comment, marked as answer]
> If two events on the same object happen very close together it is possible that the "sequencer" field could be out of order.

Locks prevent concurrent object mutations, but as anything distributed it requires your clocks to be strictly synchronized.

**Reply by mctoohey** [reply]
Thank you for your response. I guess my concern is that because the timestamp of the event is taken after the lock is released. Even if clocks are perfectly in sync. There is a still a race condition between when the lock is released and when the timestamp is taken.

It's a shame MinIO does not provide the same gaurentee as S3 in this regard. Perhaps this difference could be documented in the MinIO [Bucket Notifications](https://min.io/docs/minio/linux/administration/monitoring/bucket-notifications.html) documentation.

**Reply by klauspost** [reply]
I don't see why it would be a major issue to do that.

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/pull/18285 - here you go.
